### ISSUE TITLE
Game crashes when changing screens.

### ISSUE THREAD

**Issue by PanosPetras** [PanosPetras/The_Great_War#5]
The game now crashes when I press a button that changes screens. I noticed this behavior since commit `f2c7882`. I have managed to track it down to the `MainWindow::ChangeScreen` function. Exception thrown:

```
Exception thrown: read access violation.
drawable._Mypair.**_Myval2** was 0xFFFFFFFFFFFFFFFF.
```

When I comment out the command `scr = std::move(NewScreen);` the program doesn't function as intended but it does not crash.

**Comment by TedLyngmo** [comment]
Yes, that's the bug I'm currently trying to hunt down. While changing screen, a `Button` that belonged to the previous screen still gets its `HandleInput` called - but the `Button` is already destroyed. I'm trying to see if `gdb` can help.

**Comment by PanosPetras** [comment]
This is dreadful, I didn't even know this existed. Thanks for the information, I will give it my best shot as well.

**Comment by TedLyngmo** [comment]
I can upload my debug version to my repo so you can look at it. I've replace quite a lot of raw pointers and added debugging outputs to help. I also found a missing `virtual` destructor which caused some leaks.

**Comment by PanosPetras** [comment]
I will, thank you.

**Comment by TedLyngmo** [comment]
I've put it up as a pull request. Since I've changed so much, I don't want to submit it without you taking a look at it first. :-) I'll be going to bed now but I'll continue digging a bit Tomorrow.

**Comment by TedLyngmo** [comment]
I think I've got something of a solution. I've made `MainWindow` into a singleton (not nice, but...) and added an event queue to it. Components with the potential of destroying themselves when calling their bound functions can then put the function in the `MainWindow` event queue instead. All events put in that queue will be handled _after_ `MainWindow` has called `scr->HandleInput(event);` so the components will stay alive until afterwards. I'll soon have something to push up so you can take a look at it.

**Comment by PanosPetras** [comment]
That sounds promising. I was thinking about converting the `MainWindow` to a singleton anyways since there is always just one open window. So it only makes sense. Why do you think that this is suboptimal?

**Comment by TedLyngmo** [comment]
Singletons tend to make large projects complicated and hard to maintain. As long as there's only _one_ singleton, I guess it's ok. :-) It wouldn't be hard to instead let `MainWindow` pass a reference to itself to all the `HandleInput` functions, but that's also a bit of a drag.

**Comment by PanosPetras** [comment]
This issue was patched by commit `eba9a61`.
